spl/labels.py

```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
         return list(application_numbers)
 
     def __get_label_text(self, set_id, bs_content):
-        def get_xml_text(navigable_title):
-            # Process the label text
-            label_text = (
-                re.sub(r"\n{3,}", "\n\n", navigable_title.get_text())
-                # Remove initial / trailing white space
-                .lstrip().rstrip()
-            )
-            return label_text
+        def get_xml_text(text):
+            # Process the label text; normalize remove nbsp
+            text = unicodedata.normalize("NFKC", text.lstrip().rstrip())
+            text = re.sub(r"\n\s*\n", "\n", text)
+            return text
+
+        def get_xml_title(text):
+            # strip extra \n, \t, and spaces in titles, but keep number
+            text = get_xml_text(text)
+            text = re.sub(r"\.\s+", ".", text)
+            text = re.sub(r"\s+", " ", text)
+            return text
 
         # Get and process all "title" tags
         titles = bs_content.find_all("title")
         labels = []
 
-        for title in titles:
-            if isinstance(title, Tag):
-                for section_title in SplHistoricalLabels.LABEL_SECTIONS:
-                    if section_title in title.text:
-                        # Match the title tags in the sub-section
-                        sub_titles = title.parent.find_all("title")
-                        sub_sections = []
-                        for sub_title in sub_titles:
-                            if isinstance(sub_title, Tag):
-                                for (
-                                    sub_section_title
-                                ) in SplHistoricalLabels.LABEL_SECTIONS[
-                                    section_title
-                                ]:
-                                    if sub_section_title in sub_title.text:
-                                        # Extract the sub-title text
-                                        sub_sections.append(
-                                            {
-                                                "name": sub_section_title,
-                                                "text": get_xml_text(
-                                                    sub_title.parent
-                                                ),
-                                            }
-                                        )
-                                        # Remove the sub title from the main title
-                                        sub_title.parent.decompose()
-
-                        # Add the section text, along with the sub-sections
+        i = 0
+        while i < len(titles):
```

Comment:
For title in titles?

Reply:
I understand pythonic loops are cleaner, but in this case I am not doing it with pythonic loops since I didn't want to recursively find subtitles.  Finding sibling and child titles can be hard since some subtitles are nested in multiple sibling and multiple childs tags with no uniform tags to identify children.  It is much easier to grab all titles within the parent of the title and recognized that anything that is not the title is a subtitle. 
`t1 [t2 s1 s2 s3] t3`
To grab subtitle s1 s2 s3, it is best to just get parent of t2 and get all titles within that parent.  But if you are doing it this way you can't just go through all titles from t1 t2 s1 s2 s3 t3, or else you can tell what is subtitle or what is title.  An alternative is to test if each title is subtitle or not, but that would require traversing parents, which is another complex endeavor.

Reply:
Thanks for the explanation.

```
 since I didn't want to recursively find subtitles
```
In both cases (the old and new code), the find is not 'recursive' per se. It is as you have described here (grab all titles under the section title matched). I am not questioning the matching, just the use of indices for access which makes the code look clunky. If excluding the first match, it could be:

```
for subtitle in subtitles[1:]
```

I'll update it accordingly.